jobrunner/run.py

```diff
@@ -99,6 +118,45 @@ def handle_pending_job(job):
                 mark_job_as_running(job)
 
 
+def handle_pending_job_api(job, api):
+    if job.cancelled:
+        # Mark the job as running and then immediately invoke
+        # `handle_running_job` to deal with the cancellation. This slightly
+        # counterintuitive appraoch allows us to keep a simple, consistent set
+        # of state transitions and to consolidate all the kill/cleanup code
+        # together. It also means that there aren't edge cases where we could
+        # lose track of jobs completely after losing database state
+        mark_job_as_running(job)
+        handle_running_job_api(job, api)
+        return
+
+    awaited_states = get_states_of_awaited_jobs(job)
+    if State.FAILED in awaited_states:
+        mark_job_as_failed(
+            job, "Not starting as dependency failed", code=StatusCode.DEPENDENCY_FAILED
+        )
+    elif any(state != State.SUCCEEDED for state in awaited_states):
+        set_message(
+            job, "Waiting on dependencies", code=StatusCode.WAITING_ON_DEPENDENCIES
+        )
+    else:
+        not_started_reason = get_reason_job_not_started(job)
+        if not_started_reason:
+            set_message(job, not_started_reason, code=StatusCode.WAITING_ON_WORKERS)
+        else:
+            try:
+                set_message(job, "Preparing")
+                api.run(job_to_job_definition(job))
+                mark_job_as_running(job)
```

Comment:
Sorry, I think Dave is right. This suggestion of mine has introduced a bug where we mark jobs as failed because we encountered an error marking them as running.

Reply:
This needs to go back into an `else` clause.

Reply:
Ah, yes, will do